pycaw/magic.py

```diff
@@ -0,0 +1,812 @@
+        """Is fired, when the audio session volume/mute changed."""
+        log.debug(f"OnSimpleVolumeChanged: {self.app_exe} "
+                  f"Vol: ~{new_volume} Mute: {new_mute}")
+        # only make callbacks and update internal volume and mute
+        # if someone is listening
+
+        # TODO:
+        # Now callbacks wich are made by MagicApp or MagicSession
+        # get filtered out -> How about passing (new_volume, self_changed)
+        # self_changed will be BOLL:
+        # True when event_context.contents matches for example:
+        # self.magic_app.guid.contents
+        # ----
+        # or implement a advanced callback wich will pass this?
+        # ----
+        # or use function.__code__.co_argcount and depending if
+        # co_argcount == 1 -> make call only when new
+        # co_argcount == 2 -> make call always and pass
+        # (new_volume, self_changed)
+        # ^
+        # | This way above seems dirty... maybe an explicit flag like:
+        # MagicApp(..., always_callback=True) is better?
+        # apply same to MagicSession
+
+        # check old volume vs new:
+        if self._activated and self.volume != new_volume:
+            # self.volume will keep the none state
+            # until self._activate()
+            self.volume = new_volume
+
+            # get the guid of the 'changer'
+            changer_guid = event_context.contents
+
+            # send callbacks, if defined callback exists
+            # and if the volume is new:
+            if (self.magic_app and
+                    changer_guid != self.magic_app.guid.contents):
+
+                if self.magic_app.volume_callback:
+                    self.magic_app.volume_callback(new_volume)
+
+            if (self.magic_session and
+                    changer_guid != self.magic_session.guid.contents):
+
+                if self.magic_session.volume_callback:
+                    self.magic_session.volume_callback(new_volume)
+
+        # check old mute vs new:
+        elif self._activated and self.mute != new_mute:
+            # self.mute will keep the none state
+            # until self._activate()
+            self.mute = new_mute
+
+            # get the guid of the 'changer'
+            changer_guid = event_context.contents
+
+            # send callbacks, if defined callback exists
+            # and if the mute is new:
+            if (self.magic_app and
+                    changer_guid != self.magic_app.guid.contents and
+                    self.magic_app.mute_callback):
+
+                self.magic_app.mute_callback(new_mute)
+
+            if (self.magic_session and
+                    changer_guid != self.magic_session.guid.contents and
+                    self.magic_session.mute_callback):
+
+                self.magic_session.mute_callback(new_mute)
+
+    def OnStateChanged(self, new_state_id):
+        """Is fired, when the audio session state changed."""
+        if new_state_id == 2:
```

Comment:
Could theses states mapped to some enum so it doesn't seems to be a magic value?

Reply:
Great ^^ 🙌👍